LL1 validator

AksharP5 opened the pull request:
Rough Outline of the process: 
- [x] Grammar
    - [x] Grammar Tests
- [x] LL1Wrapper
    - [x] LL1Wrapper Tests
- [x] SpecGrammar
    - [x] SpecGrammar Tests
- [x] LL1Checker
    - [x] LL1Checker Tests
- [x] LL1Validator
    - [x] LL1Validator Tests

```
feat: add LL1Validator

Add check_ll1 method to verify ll1ness, and add helper Grammar and Wrapper classes. 

---

* Closes #18 

---
```

AksharP5 commented:
Something that has taken a bit to think about is the role of epsilon and EOF in the SpecGrammar, I was debating whether to have the SpecGrammar handle replacing empty rules with a wrapped epsilon rather than just wrapping them as they are, however I believe that we should leave that up to the LL1Checker to figure it out, as this leaves the SpecGrammar to be usable by other parts, and abides by the single responsibility principle, where the LL1Checker handles everything related to LL1, so I am documenting my decision for those things, as it may be changed in the future.

AksharP5 commented:
@StoneyJackson 

Something that has come up regarding the SpecGrammar handling a rule with a separator, currently when the SpecGrammar converts the syntactic_spec to a Grammar, the separator is being left out, if there is one, as the wrapper only takes two arguments, the nonterminal, and the specObject. After this, there is no indication for the LL1Checker to know if there was a separator in one of the rules, as the Wrapped items only have a name and object attribute. I am not sure exactly how to handle this case, should there just be another field for the LL1Wrapper for a separator? If you have any ideas, that would be great!

AksharP5 commented:
I think just wrapping it similar to any other object, and having addRule() accept a third parameter for the separator if there is one, then leaving the logic to handle that to the LL1Checker will be the best approach

AksharP5 commented:
@StoneyJackson 

For the epsilon and EOF, currently both of these are using the ```is``` and ```is not``` for comparisons, as opposed to ```==``` or ```!=```, the only thing is that this means that two different epsilon and EOF objects are not equal if they are not the same instance. I am not sure if this should be the case? I believe it shouldn't matter if we use different instances of the objects as long as their names are equal, is there a better way to test for equality on these items? The main objective of epsilon and EOF are for the FIRST and FOLLOW sets, which will then be utilized in the parsing table, the context in which epsilon will likely be used is if there is an empty rule, which will then likely get replaced by epsilon in the LL1Checker, and EOF will get added to the necessary FOLLOW sets. Is there something that I am overlooking if we simply just compare the names for these, and if they are the same, they are equal, as opposed to having to be the same instance, even if their objects are None?

StoneyJackson commented:
@AksharP5 Since you are using "" for epsilon and the actual EOF char(26) there is no risk of the user conflicting with it. So you should be able to safely use == and != instead of `is`.

My worry was that if you used "EPSILON" and "EOF", then there was the risk that the user couldn't use these names for terminals in their BNF. Using `is` would allow us to define a singleton for epsilon and eof, and then it wouldn't matter what their contents were. But no worries. Use "" and char(26) with == and !=. Should be fine.

AksharP5 commented:
@StoneyJackson 

Another thing that has come up is the addRule() method for the SpecGrammar, this currently expects all parameters that are passed to be LL1Wrapper objects, however there is no check for this currently. I was not sure if users will be calling this object and passing it invalid parameters, and if it should try and wrap these if possible, or just throw an error? Currently the addRule() method is not particularly designed to be called from the outside with non LL1Wrapper parameters, it is mostly called from inner methods when processing the passed SyntacticSpec.

StoneyJackson commented:
@AksharP5 Only the validator should be using SpecGrammar. So you can either have the validator wrap objects, or have the SpecGrammar wrap them in addRule(). I think the latter is probably better.

StoneyJackson commented on src/plcc/load_spec/validate_spec/validate_syntactic_spec/Grammar_test.py:
```diff
@@ -0,0 +1,86 @@
+
+def test_add_multiple_nonterminals(grammar, nonterminal, terminal):
+    grammar.addRule(nonterminal, [terminal])
+    diffNonterminal = getDiffNonterminal()
```
"diff" has a "standard" meaning in CS. It refers to a the difference between two things (usual text); and/or the tool that produces that diff.

How about "another"?

StoneyJackson commented on src/plcc/load_spec/validate_spec/validate_syntactic_spec/Grammar.py:
```diff
@@ -0,0 +1,61 @@
+    def getStartSymbol(self) -> str:
+        return self.startSymbol
+
+    def isTerminal(self, object: str) -> bool:
```
I'm worried that the knowledge of PLCC's syntactic rules are leaking out into this class. This knowledge belongs to the code that parses and/or the validator. But can we avoid this knowledge here? I think so...

How about this, any symbol not defined as a nonterminal must be a terminal.

```
def test_only_LHS_symbols_are_nonterminals():
    g = Grammar()
    g.addRule("a", ["b", "c"])
    g.addRule("b", [])
    assert g.isTerminal("a") == True
    assert g.isTerminal("b") == False
    assert g.isTerminal("c") == True
```

AksharP5 replied:
@StoneyJackson 

Are you suggesting that the isTerminal() function simply returns true if isNonterminal() is false for the provided object? Something like this?:
```
def isTerminal(self, object: str) -> bool: 
    if isNonterminal(object):
        return False 
    return True
```

Would you also want me to change the isNonterminal() function, and if so how else could we check for this? Would simply checking the first character being lowercase work?

Also could you please explain the test you have provided? I am not sure how ```b``` isn't a terminal, but ```a``` is?

StoneyJackson replied:
Yes... g.isTerminal(x) == not g.isNonterminal(x)

A symbol is a nonterminal if it appears on the LHS of a rule. Otherwise it is a terminal.

AksharP5 replied:
I understand that, however the part that left me confused was that 'a' appears on the LHS, but in the test it is True for being a Terminal, but I think I understand the main point. Does this mean that the nonterminals and terminals don't have a specific format that needs to be followed for the Grammar class? Any symbols are valid in a nonterminal and terminal? This should be addressed now, if you want to take another look.

StoneyJackson replied:
Oops... yes, that's a bug in my test. 'a' should be a terminal because it is on the LHS of a rule.

The format of a terminal or nonterminal is a parser thing. If you encode it here, then there is duplication, and if it ever changes, we have to change it in both the parser and in your checker. One solution is to call a parser function the checks if a symbol is a terminal or nonterminal based on its syntax. The other solution is to use the internal data structure to encode this information. We're doing the later. Terminals and nonterminals are parsed into object that encode what they are.

Now when we get to your Grammar, we are translating one internal representation into another. Your Grammar just needs the calling code to tell it which symbols are terminals and which are nonterminal. It doesn't care what their format is; just as long as they are unique. In fact, your Grammar probably doesn't even care if they are strings. They just need to be objects that work with sets and dictionaries. And since it's possible to complete decouple from the parser/format of the symbols, we should.

I'll try to take another look later today.

StoneyJackson commented on src/plcc/load_spec/validate_spec/validate_syntactic_spec/Grammar.py:
```diff
@@ -0,0 +1,61 @@
+
+    def addRule(self, nonterminal: str, form: list[str]):
+        self._checkParametersForErrors(nonterminal, form)
+        self._handleRule(nonterminal, form)
```
`_handleRule` is vague. What does `handle` mean?

StoneyJackson reviewed (changes requested):
I've reviewed Grammar.py and Grammar_test.py.

I'll just keep coming back to this and reviewing more when I have time. Don't wait though. Address whatever I've commented on as we go. Eventually, hopefully, this will settle down and we'll be ready to merge. But please don't get discouraged. It's just part of the process :)

AksharP5 commented:
@StoneyJackson 

For the Ll1Checker, there is currently an errorList that doesn't actually append anything as of now, is the idea to have the checking parsing table conflicts function append an error if the grammar is not LL1, or should it return True or False? I am not sure how the errorList will be used, as I can't think of errors that should be appearing in the LL1Checker, as everything should be caught beforehand, and currently everything in the class is just used to help build the parsing table. Not sure if we should remove the errorList, or keep it?

StoneyJackson commented on src/plcc/load_spec/validate_spec/validate_syntactic_spec/Grammar.py:
```diff
@@ -0,0 +1,56 @@
+from .errors import InvalidNonterminalError, InvalidFormError
```
Please create a subdirectory for checkLL1, and move everything related to it into that subdirectory.

StoneyJackson commented on src/plcc/load_spec/validate_spec/validate_syntactic_spec/Grammar_test.py:
```diff
@@ -0,0 +1,88 @@
+    assert len(grammar.getNonterminals()) == 1
+    assert nonterminal in grammar.getNonterminals()
+
+def test_invalid_nonterminal_throws_invalid_nonterminal_error(grammar, terminal):
```
In Grammar's universe, a valid symbol (terminal or nonterminal) is any object that words with Python's set data structure.

Otherwise it's impossible to have an invalid terminal or nonterminal. Any symbol that appears on the LHS of a rule is a nonterminal, all other symbols are terminals.

While you are at it, try to generalize the type of symbols to any object that works with sets. If there isn't a single type that does this, then generalize to object (I think that is Any in the type hints) and then check that they have the necessary methods (__eq__ and __hash__ ?). And that should be all Grammar cares about.

StoneyJackson commented on src/plcc/load_spec/validate_spec/validate_syntactic_spec/LL1Checker_test.py:
```diff
@@ -0,0 +1,93 @@
+
+def test_first_set_terminal():
+    checker = setupChecker(["<exp> ::= VAR"])
+    assert checker.firstSets["exp"] == {"VAR"}
```
Tests for LL1Checker should be of the form...

"Here is a grammar that we know is not in LL1, when we run our checker on it, it returns an error."

and 

"Here is a grammar that we know is in LL1, when we run our checker on it, it returns no errors."

And we can try different grammars that might give the checker troubles.

----

This is a test for the construction of firstSets, which should not be part of LL1Checker's observable interface (clients of LL1Checker does not and should not care about firstSets).

I think you need a separate module for FirstSets, and such tests should move to it.

Looking ahead, the same is probably true for FollowSets.

StoneyJackson commented:
Is there an entry-point function for this? I'm looking for a function like:

```python
from LL1 import check_ll1
errors = check_ll1(syntactic_spec)
```

Also, I would like to stick to the naming convention that a module/package that provides functionality is a verb-phrase. A module/package that provides a structure for storing/organizing data is a noun-phrase. For us, most of the time we are implementing functionality, so most of our modules/packages will be noun-phrases.

Also, every module/package that provides functionality should provide a top-level, entry-point function. The idea is that the caller should not care about what objects are created internally to implement the functionality. These top-level, entry-point functions are the interfaces for their components; they provide the observable functionality from outside the component, and tests should be written against these functions. This will make tests less brittle as they know as little as possible about the underlying implementation.

StoneyJackson commented:
I'm unclear on who is doing what at this point. Are others now contributing to this branch?

AksharP5 commented:
An issue of left-recursion was brought up, if you provide a left-recursive grammar and attempt to generate a FIRST set, it causes a recursion error, this will need to be resolved, it should be a similar resolution to the issue handled in #47, where the left-recursive rule should be split into two rules to to handle recursive and non-recursive parts, where the new nonterminal will handle the recursion.

AksharP5 commented:
> I'm unclear on who is doing what at this point. Are others now contributing to this branch?

Nobody else should be contributing to this branch, as far as I am aware, is there something I missed?

AksharP5 commented:
> Is there an entry-point function for this? I'm looking for a function like:
> 
> ```python
> from LL1 import check_ll1
> errors = check_ll1(syntactic_spec)
> ```
> 
> Also, I would like to stick to the naming convention that a module/package that provides functionality is a verb-phrase. A module/package that provides a structure for storing/organizing data is a noun-phrase. For us, most of the time we are implementing functionality, so most of our modules/packages will be noun-phrases.
> 
> Also, every module/package that provides functionality should provide a top-level, entry-point function. The idea is that the caller should not care about what objects are created internally to implement the functionality. These top-level, entry-point functions are the interfaces for their components; they provide the observable functionality from outside the component, and tests should be written against these functions. This will make tests less brittle as they know as little as possible about the underlying implementation.

The entry-point function would be from LL1Validator, its called checkLL1(). Also, how should it look like for providing a top-level function for something like Grammar, where there are getters that need to be called upon, it would likely just call it off the function itself, or off the variable that holds it, correct? Something like: 
```
grammar = generateEmptyGrammar()
grammar.getStartSymbol()
```

Would the top-level function being just an empty Grammar object be fine, as users can use the internal addRule method from the returned Grammar? Something like:
```
def generateEmptyGrammar():
    return Grammar()

def test_add_rule():
grammar = generateEmptyGrammar()
grammar.addRule(nonterminal, form)
```

StoneyJackson commented:
That's fine for Grammar.

Just to be clear Grammar is not the interface of the check_ll1 package. The check_ll1 function of the check_ll1 module is...

```
check_ll1/
    __init__.py   <-- from .check_ll1 import check_ll1
    check_ll1.py    <-- def check_ll1(syntactic_spec): ...
```

When someone calls check_ll1, it will need to build a Grammar from the SyntacticSpec and then check it. Probably both of these procedures are separate sub-functions and would be defined in separate modules and/or packages.

If another component wanted to use Grammar, it would have to reach inside the check_ll1 package to a subpackage or submodule where Grammar and its methods are the interface.

StoneyJackson commented:
@AksharP5 I started to review this today. But I think you're still in the middle of changes. Ping me on discord when you're ready for a review.

StoneyJackson commented on src/plcc/load_spec/validate_spec/validate_syntactic_spec/LL1/LL1Validator.py:
```diff
@@ -0,0 +1,13 @@
+from .SpecGrammar import createSpecGrammar
+
+def checkLL1(syntactic_spec):
+    LL1Validator(syntactic_spec).validate()
```
Needs a return.

StoneyJackson commented on src/plcc/load_spec/validate_spec/validate_syntactic_spec/LL1/LL1Validator.py:
```diff
@@ -0,0 +1,13 @@
+from .LL1Checker import check
+from .SpecGrammar import createSpecGrammar
+
+def checkLL1(syntactic_spec):
```
For consistency, let's use `check_ll1` validator.

We use camelCasing for OO, and snake_casing otherwise.

StoneyJackson commented on src/plcc/load_spec/validate_spec/validate_syntactic_spec/LL1/SpecGrammar.py:
```diff
@@ -0,0 +1,99 @@
+from .LL1Wrapper import wrapLL1, LL1Wrapper
+from .errors import InvalidSymbolError, InvalidSyntacticSpecError
+
+def createSpecGrammar(syntacticSpec: SyntacticSpec):
```
snake_casing: create_grammar and syntactic_spec (but not for the type SyntacticSpec).

StoneyJackson replied:
I'm going to stop making this comment, but you should apply it throughout.

StoneyJackson commented on src/plcc/load_spec/validate_spec/validate_syntactic_spec/errors.py:
```diff
@@ -2,7 +2,7 @@
 
 @dataclass
-class ValidationError:
+class ValidationError(Exception):
```
Are ValidationErrors actually ever raised as an Exception through Python's exception handling mechanism? If not, I think we should not make them inherit from Exception.

AksharP5 replied:
@StoneyJackson I made the ValidationError class inherit from Exception for testing purposes, as ```with raises``` expects a BaseException type, is it fine to leave it like this, or should I remove it? If I remove it, I am not sure how to test the errors, as there is no errorList for many of the sections, such as Grammar, which throws a few errors.

StoneyJackson replied:
We shouldn't extend Exception "for testing purposes". We should if we raise it as an exception.

When should we raise an exception? When the current algorithm cannot continue as normal because of the detected error.

Should a validator raise an exception? The purpose of a validator is to detect errors. If it is possible to detect multiple errors, it should do so. When it is done finding all the problems it can, it should return them as a list of errors. The only reason a validator would raise an exception is if there is something that would prevent it from doing its job.

Assuming a ValidationError represents an error detected by a validator (and not an error with the validator itself) it should not necessarily be an Exception.

One last thought. There is a chance later down the road that we actually do want to raise them as exceptions. But that's not clear at this point.

AksharP5 replied:
@StoneyJackson ValidationErrors are extending Exceptions in different parts of the code as well, this may need to be added onto #38, and be dealt with in a bigger sense. But in the sense of this PR, there are errors that do not allow the LL1check to progress, would you like these extend Exception rather than ValidationError? Also, for the errors that don't prevent progress, would you like me to add an errorList that gets returned? I am thinking that the same errorList would go through each step and be extended, if needed, and that would be what is returned at the very end?

StoneyJackson replied:
I see... there should be a separation between a ValidationError and an exception that is raised when something breaks a validator.

I suggest that you rename all the exceptions with `Exception` at the end of their name, and have them extend `Exception` directly (and not `ValidationError`).

Then, for those errors that are not raised and represent errors detected by a validator, have them extend ValidatorError.

We'll worry about what the ValidatorError extends later.

StoneyJackson reviewed (changes requested):
I think 95% of my requests are cosmetic. So I think we are very close. In fact this might be it :)

AksharP5 commented:
@StoneyJackson 

The issue of left-recursion is also now being rejected in the FIRST set generation.

AksharP5 committed:
feat: add LL1Validator

Add check_ll1 method to verify ll1ness, and add helper Grammar and Wrapper classes.

---

* Closes #18
* Closes #48

---

connor-a-francis committed:
feat: replace repeating with standard rules

StoneyJackson committed:
Merge pull request #47 from ourPLCC/32-transform-repeating-rules-into-standard-bnf-rules

feat: resolve repeating rules into standard rules

StoneyJackson commented:
> @StoneyJackson
> 
> The issue of left-recursion is also now being rejected in the FIRST set generation.

What do you mean "is now being rejected" ?

KyleAlme committed:
feat: add parsing table checker

---

* Closes #36

---

michaelBanerjee committed:
feat: add build_parsing_table

AksharP5 commented:
> > @StoneyJackson
> > The issue of left-recursion is also now being rejected in the FIRST set generation.
> 
> What do you mean "is now being rejected" ?

There is an error (now exception) being thrown, when left-recursion is detected.

AksharP5 committed:
update errors to exceptions

StoneyJackson committed:
Merge pull request #49 from ourPLCC/36-check-parsing-table-for-ll1-ness

Check parsing table for ll1 ness

StoneyJackson committed:
Merge pull request #46 from ourPLCC/michaelbanerjee/create-parsing-33

Create Parsing Table

StoneyJackson committed:
refactor: move build_parsing_table int LL1

StoneyJackson committed:
refactor: rename LL1 and sub units for consistency

StoneyJackson commented on src/plcc/load_spec/validate_spec/validate_syntactic_spec/validate_ll1/generate_follow_sets.py:
```diff
@@ -0,0 +1,70 @@
+        self.grammar = grammar
+        self.firstSets = generate_first_sets(grammar)
+        self.followSets = defaultdict(set)
+        self.FirstSetGenerator = FirstSetGenerator(grammar)
```
We already generated the first sets, why do we need a FirstSetGenerator?

StoneyJackson commented on src/plcc/load_spec/validate_spec/validate_syntactic_spec/validate_ll1/generate_follow_sets.py:
```diff
@@ -0,0 +1,70 @@
+        return updated
+
+    def _addFirstOfNextSymbol(self, nextSymbol, nonterminal):
+        nextFirst = self.FirstSetGenerator._computeFirst(nextSymbol) - {self.grammar.getEpsilon().name}
```
Here is the use of FirstSetGenerator. But we're violating its encapsulation to call _computerFirst(). Shouldn't firstSets have what we need?

I tried replacing this call with `self.firstSets[nextSymbol]` but it doesn't pass the tests. But I don't understand why.

I would really like to remove this strange dependency. Any ideas?

AksharP5 replied:
@StoneyJackson I also ran into this same issue when refactoring these into separate files, and I believe that the tests do not pass strictly off the self.firstSets, because the firstSets only holds the FIRST sets for the nonterminals, but the FOLLOW set generation needs to utilize the FIRST sets for all the symbols, even terminals, which is done by using the _computeFirst() function. The way that this could be fixed is if the FirstSetGenerator's firstSet held the FIRST set for all symbols?

AksharP5 committed:
remove FirstSetGenerator from generate_follow_sets

StoneyJackson commented:
@michaelBanerjee and @AksharP5 

Grammar and the table builder use different structures to represent grammar rules.

* Grammar (really SpecGrammar) uses dict[str, list[list[object]]]
* Table builder uses dict[str, list[str]]

Ideally, the type should be dict[object, list[list[object]]], where object is an immutable object that with __hash__ and __eq__ methods.

Is it possible to make this adjustment?

If not, we would need to write code to convert one to the other, and I fear this is making a more brittle system because each places slightly different set of constraints on symbols.

I tried to make quick obvious changes to the table builder. But those are not enough; it requires more thought and effort which is probably better performed by the person who built it.

AksharP5 committed:
setup validate_ll1

AksharP5 committed:
update generate_first_sets to include all first sets, including rules

AksharP5 commented:
> @michaelBanerjee and @AksharP5
> 
> Grammar and the table builder use different structures to represent grammar rules.
> 
> * Grammar (really SpecGrammar) uses dict[str, list[list[object]]]
> * Table builder uses dict[str, list[str]]
> 
> Ideally, the type should be dict[object, list[list[object]]], where object is an immutable object that with **hash** and **eq** methods.
> 
> Is it possible to make this adjustment?
> 
> If not, we would need to write code to convert one to the other, and I fear this is making a more brittle system because each places slightly different set of constraints on symbols.
> 
> I tried to make quick obvious changes to the table builder. But those are not enough; it requires more thought and effort which is probably better performed by the person who built it.

There is also a difference with the first sets being used in the tests, I tried taking a look and making changes, but I am not very familiar with the algorithm for building the parsing table, I think leaving it to @michaelBanerjee would be best for this step. The current firstSets are only generated for the nonterminals and terminals, the current build_parsing_table tests expect firstSets to be generated for all the rules, which is not the case. I went ahead and made changes to the firstSet generation to include the first sets for all rules, as expected for the build_parsing_table tests, but I will leave updating the algorithm to @michaelBanerjee. Also, there is a test for left-recursion, which should be removed, as it will throw an exception when found, there is no need to test for it, as we will never be able to duplicate it.

StoneyJackson commented:
@AksharP5 Thanks for the close look!

AksharP5 committed:
rename test

StoneyJackson committed:
fix(Grammar): when terminal found to be nonterminal remove it from terminals)

StoneyJackson committed:
fix(Grammar): any set-compatible object is a valid nonterminal

StoneyJackson committed:
refactor(Grammar): let set detect invalid symbols

If Grammar is passed an invalid symbol, Python's set will detect
it and raise a TypeError. We don't need an extra layer of code
to detect and translate these errors into a different exception.

Note, we would need this extra layer of code if this was a user
error that we were trying to detect and report to the user. But
this is an internal programmer error that should be detected
and reported by unit tests.

StoneyJackson committed:
refactor(Grammar): let set detect invalid symbols

If Grammar is passed an invalid symbol, Python's set will detect
it and raise a TypeError. We don't need an extra layer of code
to detect and translate these errors into a different exception.

Note, we would need this extra layer of code if this was a user
error that we were trying to detect and report to the user. But
this is an internal programmer error that should be detected
and reported by unit tests.

StoneyJackson committed:
refactor(Grammar): any iterable is a valid form

The second parameter to addRule() must be an iterable. With this change,
a str passed as a second parameter would result in each character being
treated as a symbol. If this was not the intent of the programmer,
then this is an error. However, this error should be detected by unit
tests.

StoneyJackson committed:
refactor(Grammar): simplify and clarify

* Replace str with object in all public methods.
* Remove nonterminals set and use keys from rules instead.
* isTerminal(s) is true only if s has been seen
     and is not a nonterminal.
* General cleanning and reorganizing.

StoneyJackson committed:
refactor(Grammar): remove generate_grammar()

Grammer is a data structure (and not a functional unit),
and generate_grammar does not provide any additional functionality.
So delete generate_grammar().

StoneyJackson committed:
refactor: use "build"

We had too many names for the same thing:

* build
* generate
* create

Let's use the shorter: "build".

michaelBanerjee committed:
update build_parsing_table() to use getRules()

michaelBanerjee commented:
@StoneyJackson 
build_parsing_table() now uses dict[str, list[list[LL1wrapper]] as it's rules parameter.

Correct me if I am missing something, but I think the only way not to convert a list of LL1Wrapper objects to a string (which I have implemented) would be to change the keys of first and follow sets, as well as the values of getRules() to a an immutable object class. 
These would have to be the same type since values given from rules are needed to lookup the firstSets of productions. LL1Wrapper cannot be used for this I believe, since it only represent a single nonterminal/terminal, and not multiple. Right now, this is handled by having a list of LL1Wrapper objects. 
This could mean either making getRules() return dict[str, list[str]], or using a new immutable object that would likely contain LL1Wrapper objects.

This is out of the scope of build_parsing_table(), and I'm not sure if it's for the best for the other functions which would be affected by this change. 

However, if these changes are implemented it would be possible to not convert these lists of LL1Wrappers to strings.

Side note: The test for left-recursion has been removed.

StoneyJackson committed:
refactor(build_first_sets): remove memoFirst

memoFirst is an optimization. We don't know that we need it.
So it's premature optimization. And Donald Knuth says,
"premature optimization is the root of all evil." So it adds
obvious complexity without a measurable, significant gain.

StoneyJackson committed:
refactor(build_first_sets): remove check for left recursion

This clutters the algorithm. We will add this as a separate module.

StoneyJackson committed:
refactor: clean build_first_sets and build_follow_sets

StoneyJackson committed:
refactor(check_ll1): integrating

* Unify types across components.
* Remove dependencies and simplify when possible.

StoneyJackson commented:
Almost there. Need at pass over the code at least once more.

When I merge, I don't plan to squash. Too much code with too many author's is involved. I'd rather keep the history so everyone gets credit for the work they did.

StoneyJackson committed:
refactor(build_parsing_table): clean

StoneyJackson committed:
refactor(build_follow_sets): clean

StoneyJackson commented:
Test case needed for build_follow_sets: When B can derive epsilon (see Figure 2 in https://pages.cs.wisc.edu/~hasti/cs536/readings/Topdown.html#follow).

StoneyJackson committed:
test(build_bollow_sets): up coverage, add acceptance tests

* Add a test to cover case when follows of subsequent
  symbols all contain epsilon, and so follows then
  includes parent's follows.

* In code, removed a test for symbol being epsilon.
  This shouldn't be possible because Grammar rules
  do not contian a symbol for epsilon; epsilon is
  represented by an empty production. However, a
  symbol IS still used for first sets, and Grammar
  does hold a symbol for use in first sets.

StoneyJackson commented:
🎉 🥁 🎸
